### PR TITLE
Support terms filtering

### PR DESCRIPTION
**Pull request by alexksikes**
This adds a new feature to the Term Vectors API which allows for filtering of
terms based on their tf-idf scores. With `dfs` option on, this could be useful
for finding out a good characteristic vector of a document or a set of documents.
The parameters are similar to the ones used in the MLT Query.

### CODE REVIEW

File: src/main/java/org/elasticsearch/action/termvectors/TermVectorsFilter.java [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+            // one queue per field name
+            ScoreTermsQueue queue = new ScoreTermsQueue(Math.min(maxNumTerms, (int) terms.size()));
+
+            // select terms with highest tf-idf
+            termsEnum = terms.iterator(termsEnum);
+            topLevelTermsEnum = topLevelTerms.iterator(topLevelTermsEnum);
+            while (termsEnum.next() != null) {
+                BytesRef termBytesRef = termsEnum.term();
+                topLevelTermsEnum.seekExact(termBytesRef);
```
Comment (jpountz) [inline src/main/java/org/elasticsearch/action/termvectors/TermVectorsFilter.java]: Is it ok to ignore the return value?

File: src/main/java/org/elasticsearch/action/termvectors/TermVectorsFilter.java [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        // index update problem?
+        if (docFreq == 0) {
+            return false;
+        }
+        // filter out words based on length
+        int len = word.length();
+        if (minWordLength > 0 && len < minWordLength) {
+            return false;
+        }
+        if (maxWordLength > 0 && len > maxWordLength) {
```
Comment (jpountz) [inline src/main/java/org/elasticsearch/action/termvectors/TermVectorsFilter.java]: Do we actually need all these if (minDocFreq|maxWordLen|... > 0) ?

File: src/main/java/org/elasticsearch/index/termvectors/ShardTermVectorsService.java [hunk 3]
```diff
@@ -332,6 +347,10 @@ private boolean useDfs(TermVectorsRequest request) {
         return request.dfs() && (request.fieldStatistics() || request.termStatistics());
     }
 
+    private boolean useTermVectorsFilter(TermVectorsRequest request) {
+        return request.filterSettings() != null && request.fieldStatistics() && request.termStatistics();
```
Comment (jpountz) [inline src/main/java/org/elasticsearch/index/termvectors/ShardTermVectorsService.java]: If filter settings are set but statistics are not requested then we ignore the filter settings?

**Comment by jpountz** [comment]
Looks good overall, can you add some tests?

File: docs/reference/docs/termvectors.asciidoc [hunk 4]
```diff
@@ -86,6 +86,40 @@ or the field statistics of the entire index, and not just at the shard. Use it
 with caution as distributed frequencies can have a serious performance impact.
 
 [float]
+==== Terms Filtering coming[2.0]
+
+With the parameter `filter`, the terms returned could also be filtered based
```
Comment (clintongormley) [inline docs/reference/docs/termvectors.asciidoc]: I'm nervous about using the word `filter` here, as we try to reserve that purely for query DSL filters (although we do use it for fielddata filters as well http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/fielddata-formats.html#field-data-filtering). I'm struggling to come up with a better name though.

I'm wondering if we need this sub-level at all, or if we can just move the `max_num_terms` etc parameters up to the top level?  There doesn't seem to be any name clash (although I do like the fact that having them as sub-params indicates their action more clearly).

Comment (alexksikes) [inline docs/reference/docs/termvectors.asciidoc]: One nice thing about using `filter

Comment (alexksikes) [inline docs/reference/docs/termvectors.asciidoc]: One nice thing about using `filter` as a sub-level parameter is that `filter : {}` with no parameter will perform filtering but using the default sub parameters. Another reason would be to hide the complexity of a feature which might be used only in some very specific cases. It also allows us to expand on this sub feature, in order, for example, to allow filtering based on a script more cleanly I think. But I do agree, I am not sure if `filter` should be the proper name for this feature.

**Comment by s1monw** [comment]
@alexksikes are you picking this up again or can we close it?

**Comment by alexksikes** [comment]
Just need to add the tests for this. We would need this for Item Query also for which I wanted to write a dev issue to explain the plan.

File: src/main/java/org/elasticsearch/action/termvectors/TermVectorsFilter.java (hunk 1 above)
Comment (jpountz) [inline src/main/java/org/elasticsearch/action/termvectors/TermVectorsFilter.java]: Can you just assert that it returns true?

**Comment by jpountz** [comment]
LGTM, just left a minor comment that does not need further review. However before pushing can you make sure to settle on the naming of this feature with @clintongormley ?

**Comment by clintongormley** [comment]
I can't think of a better name for this parameter, and it fits with fielddata filtering, so I think we should just stick to `filter` here (especially in light of the fact that query DSL filters will kinda go away in the future)

**Pull request closed without merging**